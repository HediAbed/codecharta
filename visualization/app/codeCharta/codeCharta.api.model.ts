import {
	AttributeTypes,
	AttributeTypeValue,
	CodeMapNode,
	Edge,
	MarkedPackage,
	MetricStatisticsItem,
	RecursivePartial,
	Settings
} from "./codeCharta.model"

export interface ExportCCFile {
	projectName: string
	apiVersion: string
	fileChecksum: string
	nodes: CodeMapNode[]
	attributeTypes?: AttributeTypes | OldAttributeTypes
	edges?: Edge[]
	markedPackages?: MarkedPackage[]
	blacklist?: ExportBlacklistItem[]
	metricStatistics?: ExportMetricStatisticsItem[]
}

export interface ExportBlacklistItem {
	path: string
	type: ExportBlacklistType
}

export type ExportMetricStatisticsItem = MetricStatisticsItem

export enum ExportBlacklistType {
	hide = "hide",
	exclude = "exclude"
}

export enum APIVersions {
	ZERO_POINT_ONE = "0.1",
	ONE_POINT_ZERO = "1.0",
	ONE_POINT_ONE = "1.1",
	ONE_POINT_TWO = "1.2"
}

export interface ExportScenario {
	name: string
	settings: RecursivePartial<Settings>
}

export interface OldAttributeTypes {
	nodes?: [{ [key: string]: AttributeTypeValue }?]
	edges?: [{ [key: string]: AttributeTypeValue }?]
}
