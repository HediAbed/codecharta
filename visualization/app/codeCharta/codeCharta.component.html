<tool-bar-component></tool-bar-component>

<file-extension-bar-component ></file-extension-bar-component>
<div style="margin-top:1px; width: 100%; min-height: 100px; position: fixed; z-index: 10; overflow: hidden; white-space: nowrap;font-size: 0px;" >	
	<ribbon-bar-component class="ribbon-bar-component"></ribbon-bar-component >
</div>


<div id="actionButtons">
	<md-button
		ng-if="$ctrl._viewModel.focusedNodePath.length > 0"
		class="md-fab md-primary dark_shadow_opacity long-button"
		aria-label="Remove Focus on node"
		ng-click="$ctrl.removeFocusedNode()"
	>
		Show complete map
	</md-button>
	<md-button class="md-fab md-primary dark_shadow_opacity" aria-label="Fit Map to View" ng-click="$ctrl.fitMapToView()">
		<i class="fa fa-compress fa-lg"></i>
	</md-button>
</div>

<code-map-component></code-map-component>

<legend-panel-component></legend-panel-component>

<detail-panel-component></detail-panel-component>

<loading-gif-file-component></loading-gif-file-component>

<div id="mw-logo">
	<div class="logo">
		<a href="https://www.maibornwolff.de" target="_blank"><img src="assets/logo.png" alt="MaibornWolff"/></a>
		<h2>
			<span>CodeCharta {{ $ctrl._viewModel.version }}</span>
		</h2>
	</div>
</div>

<node-context-menu-component></node-context-menu-component>
