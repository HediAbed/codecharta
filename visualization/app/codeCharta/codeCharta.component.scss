@import "../mixins.scss";

code-charta-component {
	$animationDuration: 0.5;

	#mw-logo {
		.logo {
			width: 160px;
			height: 100px;
			position: absolute;
			left: 0;
			bottom: 30px;
			z-index: 29;
		}

		a,
		img {
			width: 5em;
			display: block;
			margin: auto !important;
		}

		h2 {
			text-transform: uppercase;
			color: #5a585a;
			font-family: Roboto, sans-serif;
			font-weight: 500;
			font-size: 9pt;
			text-align: center;
			margin: 0;
			z-index: -1;
		}
	}

	#title {
		position: fixed;
		z-index: 2;
		width: 100%;
	}

	.small-info {
		color: rgba(0, 0, 0, 0.38);
		font-size: 9pt;
	}

	#codeMap {
		position: fixed;
		z-index: 0;
		top: 90px;
		left: 0;
		width: 100%;
		height: 100%;
		transition: all $animationDuration + s ease;

		&.expanded {
			height: calc(100% - 170px);
		}
	}

	#actionButtons {
		position: fixed;
		bottom: 2em;
		right: 2em;
		z-index: 60;

		i.fa-folder-open {
			margin: 4px 0 0 2px;
		}

		.long-button {
			width: inherit;
			padding-left: 10px;
			padding-right: 10px;
			border-radius: 100px;
		}

		> * {
			margin-left: 4px;
			margin-right: 4px;
			z-index: 60;
		}
	}

	.dark_shadow_opacity {
		box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2) !important;
		opacity: 0.97;
	}

	.bottom_shadow {
		box-shadow: 0 4px 4px -2px rgba(0, 0, 0, 0.2) !important;
	}

	.md-button.round-button {
		margin-right: 0 !important;
		float: right;
		min-width: 36px;
		width: 36px;
		border-radius: 50%;
	}

	.ribbon-bar-component {
		z-index: 10;

		md-card,
		.section .section-header,
		.section .section-body,
		#toggle-ribbon-bar-fab {
			transition: all $animationDuration + s ease;
		}
	}

	.ribbon-bar-wrapper {
		margin-top: 1px;
		width: 100%;
		min-height: 100px;
		height: 500px;
		position: fixed;
		z-index: 10;
		overflow: hidden;
		white-space: nowrap;
		font-size: 0px;
		pointer-events: none;
	}

	.angular-green {
		color: #00bfa5;
	}

	.small-action-button {
		height: 24px;
		width: 24px;
		float: left;

		*,
		.md-fab:active {
			margin: 0;
			padding: 0;
		}

		.md-fab {
			position: relative;
			min-height: 0;
			width: 100%;
			height: 100%;
			z-index: 0;
			border-radius: 50%;

			i {
				position: absolute;
				top: 5px;
				left: 6px;
				color: white;

				&:hover {
					cursor: pointer;
					color: #eee;
				}
			}
		}
	}
}
