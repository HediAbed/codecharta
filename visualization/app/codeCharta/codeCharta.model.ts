export interface CCFile {
  map: CodeMapNode;
  settings: {
      fileSettings: RecursivePartial<Settings>
  };
  fileMeta: FileMeta;
}

export interface CodeMapNode {
    name: string;
    type: string;
    children?: CodeMapNode[];
    attributes: {
        [key: string]: number
    };
    deltas?: {
        [key: string]: number
    };
    link?: string;
    origin?: string;
    visible?: boolean;
    path?: string;
    markingColor?: string;
}

export interface FileMeta {
  fileName: string;
  apiVersion: string;
  projectName: string;
}

export interface Settings {
    fileSettings: FileSettings;
    dynamicSettings: DynamicSettings;
    appSettings: AppSettings;
}

export interface FileSettings {
    attributeTypes: AttributeTypes;
    blacklist: Array<BlacklistItem>;
    edges: Edge[];
    markedPackages: MarkedPackages[];
}

export interface DynamicSettings {
    renderMode: RenderMode; // TODO remove from settings ?
    areaMetric: string;
    heightMetric: string;
    colorMetric: string;
    focusedNodePath: string;
    searchedNodePaths: Array<string>;
    searchPattern: string;
    margin: number;
    neutralColorRange: ColorRange;
}

export interface AppSettings {
    amountOfTopLabels: number;
    scaling: Vector3d;
    camera: Vector3d;
    deltaColorFlipped: boolean;
    enableEdgeArrows: boolean;
    hideFlatBuildings: boolean;
    maximizeDetailPanel: boolean;
    invertHeight: boolean;
    dynamicMargin: boolean;
    isWhiteBackground: boolean;
    mapColors: MapColors;
    whiteColorBuildings: boolean;
}

export interface MapColors {
    positive: string;
    neutral: string;
    negative: string;
    selected: string;
    defaultC: string;
    positiveDelta: string;
    negativeDelta: string;
    base: string;
    flat: string;
    lightGrey: string;
    angularGreen: string;
    markingColors: string[];
}

export interface ColorRange {
    from: number;
    to: number;
    flipped: boolean;
}

export interface Vector3d {
    x: number;
    y: number;
    z: number;
}

export interface AttributeTypes {
    nodes?: {
        [key: string]: AttributeType
    };
    edges?: {
        [key: string]: AttributeType
    };
}

export enum AttributeType {
    absolute, relative
}


export interface Edge {
    fromNodeName: string;
    toNodeName: string;
    attributes: {
        [key: string]: number
    };
    visible?: boolean;
}

export interface BlacklistItem {
    path: string;
    type: BlacklistType;
}

export enum BlacklistType {
    hide = "hide",
    exclude = "exclude"
}

export interface MarkedPackages {
    path: string,
    color: string,
    attributes: {
        [key: string]: any
    }
}

export enum RenderMode {
  Single = "Single",
  Multiple = "Multiple",
  Delta = "Delta",
}

export interface MetricData {
    name: string;
    maxValue: number;
}

export interface Scenario {
    name: string;
    settings: RecursivePartial<Settings>;
}

export interface UrlData {
    filenames: string[];
    settings: Partial<Settings>;
}

export type RecursivePartial<T> = {
    [P in keyof T]?: RecursivePartial<T[P]>;
};
