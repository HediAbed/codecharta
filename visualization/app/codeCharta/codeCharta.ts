"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent} from "./codeCharta.component";

import angular from "angular";
import { CodeChartaService } from "./codeCharta.service";
import _ from "lodash";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
).service(
    _.camelCase(CodeChartaService.name),
    CodeChartaService
);


