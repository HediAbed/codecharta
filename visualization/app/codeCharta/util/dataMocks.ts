import { CodeMapNode, Edge, CCFile, Settings, FileSelectionState } from "../codeCharta.model"

export const VALID_NODE: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	children: [
		{
			name: "big leaf",
			type: "File",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			children: [
				{
					name: "small leaf",
					type: "File",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				}
			]
		}
	]
}

export const VALID_NODE_WITH_PATH: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	path: "/root",
	children: [
		{
			name: "big leaf",
			type: "File",
			path: "/root/big leaf",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			path: "/root/Parent Leaf",
			children: [
				{
					name: "small leaf",
					type: "File",
					path: "/root/Parent Leaf/small leaf",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					path: "/root/Parent Leaf/other small leaf",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				},
				{
					name: "empty folder",
					type: "Folder",
					path: "/root/Parent Leaf/empty folder",
					attributes: {},
					children: []
				}
			]
		}
	]
}

export const VALID_EDGES: Edge[] = [
	{
		fromNodeName: "/root/big leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	},
	{
		fromNodeName: "/root/sample1 only leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	}
]

export const VALID_EDGE: Edge = {
	fromNodeName: "/root/big leaf",
	toNodeName: "/root/Parent Leaf/small leaf",
	attributes: {
		pairingRate: 89,
		avgCommits: 34
	}
}

export const TEST_FILE_CONTENT = {
	fileName: "noFileName",
	projectName: "Sample Map",
	apiVersion: "1.1",
	nodes: [VALID_NODE]
}

export const TEST_FILE_DATA: CCFile = {
	fileMeta: {
		fileName: "file",
		projectName: "Sample Map",
		apiVersion: "1.1"
	},
	map: VALID_NODE,
	settings: {
		fileSettings: {
			edges: VALID_EDGES
		}
	}
}

export const TEST_FILE_WITH_PATHS: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		path: "/root",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				path: "/root/big leaf",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				path: "/root/Parent Leaf",
				children: [
					{
						name: "small leaf",
						type: "File",
						path: "/root/Parent Leaf/small leaf",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						path: "/root/Parent Leaf/other small leaf",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					},
					{
						name: "empty folder",
						type: "Folder",
						path: "/root/Parent Leaf/empty folder",
						attributes: {},
						children: []
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {}
	}
}

export const TEST_DELTA_MAP_A: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {}
	}
}

export const TEST_DELTA_MAP_B: CCFile = {
	fileMeta: {
		fileName: "fileB",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 20, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "additional leaf",
				type: "File",
				attributes: { rloc: 10, functions: 11, mcc: 5 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100, more: 20 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000 }
					},
					{
						name: "big leaf",
						type: "File",
						attributes: { rloc: 20, functions: 10, mcc: 1 },
						link: "http://www.google.de"
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {}
	}
}

export const TEST_FILE_DATA_DOWNLOADED = {
	apiVersion: "1.1",
	attributeTypes: undefined,
	blacklist: undefined,
	edges: [
		{ attributes: { avgCommits: 34, pairingRate: 89 }, fromNodeName: "/root/big leaf", toNodeName: "/root/Parent Leaf/small leaf" },
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/sample1 only leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		}
	],
	fileName: "file.14_12_2018.json",
	nodes: [
		{
			attributes: {},
			children: [
				{ attributes: { Functions: 10, MCC: 1, RLOC: 100 }, link: "http://www.google.de", name: "big leaf", type: "File" },
				{
					attributes: {},
					children: [
						{ attributes: { Functions: 100, MCC: 100, RLOC: 30 }, name: "small leaf", type: "File" },
						{ attributes: { Functions: 1000, MCC: 10, RLOC: 70 }, name: "other small leaf", type: "File" }
					],
					name: "Parent Leaf",
					type: "Folder"
				}
			],
			name: "root",
			type: "Folder"
		}
	],
	projectName: "Sample Map"
}

export const SETTINGS: Settings = {
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	dynamicSettings: {
		areaMetric: "rloc",
		heightMetric: "mcc",
		colorMetric: "mcc",
		focusedNodePath: null,
		searchedNodePaths: [],
		searchPattern: "",
		margin: 48,
		neutralColorRange: { flipped: false, from: 19, to: 67 }
	},
	appSettings: {
		amountOfTopLabels: 31,
		scaling: { x: 1, y: 1.8, z: 1 },
		camera: { x: 0, y: 300, z: 1000 },
		deltaColorFlipped: false,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		maximizeDetailPanel: false,
		invertHeight: true,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: true,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}
	},
	treeMapSettings: { mapSize: 500 }
}
