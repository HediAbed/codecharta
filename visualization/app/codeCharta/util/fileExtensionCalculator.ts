import { CodeMapNode, KeyValuePair } from "../codeCharta.model"
import * as d3 from "d3"
import { HierarchyNode } from "d3"
import _ from "lodash"

export interface ExtensionAttribute {
	fileExtension: string,
	absoluteMetricValue: number,
	relativeMetricValue: number,
	color: string
}

export interface MetricDistribution {
	metric: string
	distribution: ExtensionAttribute[]
}

export class FileExtensionCalculator {
	private static NO_EXTENSION = "none"
	private static distribution: MetricDistribution[]

	// TODO: this does not exclude blacklisted nodes yet
	public static getAbsoluteFileExtensionDistribution(map: CodeMapNode, metrics: string[]): MetricDistribution[] {
		this.initEmptyDistribution(metrics)

		d3.hierarchy(map)
			.leaves()
			.forEach((node: HierarchyNode<CodeMapNode>) => {
				const fileExtension: string = this.estimateFileExtension(node.data.name)
				metrics.forEach((metric: string) => {
					const metricValue: number = node.data.attributes[metric]
					const extensionPointer: ExtensionAttribute[] = this.distribution.find(x => x.metric === metric).distribution
					const matchingItem: ExtensionAttribute = extensionPointer.find(x => x.fileExtension === fileExtension)

					if (matchingItem) {
						matchingItem.absoluteMetricValue += metricValue
					} else {
						extensionPointer.push({
							fileExtension: fileExtension,
							absoluteMetricValue: metricValue,
							relativeMetricValue: null,
							color: null
						})
					}
				})
			})
		return this.distribution
	}

	public static getRelativeFileExtensionDistribution(map: CodeMapNode, metrics: string[]): MetricDistribution[] {
		const distribution: MetricDistribution[] = this.getAbsoluteFileExtensionDistribution(map, metrics)

		distribution.forEach((distribution: MetricDistribution) => {
			const sumOfAllMetricValues: number = this.getSumOfAllMetrics(distribution)
			distribution.distribution.forEach((x: ExtensionAttribute) =>
					x.relativeMetricValue = Math.round(x.absoluteMetricValue / sumOfAllMetricValues * 10000) / 100
			)
			distribution.distribution.sort((a,b) => b.absoluteMetricValue - a.absoluteMetricValue)
		})

		return distribution
	}

	private static getSumOfAllMetrics(distribution: MetricDistribution): number {
		return distribution.distribution.map(x => x.absoluteMetricValue)
			.reduce((partialSum, a) => partialSum + a)
	}

	private static initEmptyDistribution(metrics: string[]) {
		this.distribution = []

		metrics.forEach((metric: string) => {
			this.distribution.push({
				metric: metric,
				distribution: []
			})
		})
	}

	private static estimateFileExtension(fileName: string): string {
		if (fileName.includes(".")) {
			return fileName
				.split(".")
				.reverse()[0]
				.toLowerCase()
		} else {
			return FileExtensionCalculator.NO_EXTENSION
		}
	}
}
