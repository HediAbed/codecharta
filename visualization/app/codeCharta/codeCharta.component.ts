import { NameDataPair, UrlUtils } from "./util/urlUtils"
import {IHttpService, ILocationService, IRootScopeService} from "angular"
import "./codeCharta.component.scss"
import { NodeContextMenuController } from "./ui/nodeContextMenu/nodeContextMenu.component"
import { CodeMapActionsService } from "./ui/codeMap/codeMap.actions.service"
import { CodeChartaService } from "./codeCharta.service"

/**
 * This is the main controller of the CodeCharta application
 */
export class CodeChartaController {
	public viewModel = {
		version: require("../../package.json").version,
		numberOfLoadingTasks: 0
	}
	private urlUtils: UrlUtils

	/* @ngInject */
	constructor(
		//private threeOrbitControlsService: ThreeOrbitControlsService,
		private $rootScope: IRootScopeService,
		//private dialogService: DialogService,
		//private codeMapActionsService: CodeMapActionsService,
		private codeChartaService: CodeChartaService,
		private $location: ILocationService,
		private $http: IHttpService
	) {
		this.urlUtils = new UrlUtils(this.$location, this.$http)
		this.subscribeToLoadingEvents(this.$rootScope)
		this.loadFileOrSample()
	}

	public fitMapToView() {
		// this.threeOrbitControlsService.autoFitTo()
	}

	public removeFocusedNode() {
		//this.codeMapActionsService.removeFocusedNode()
	}

	public loadFileOrSample() {
		this.viewModel.numberOfLoadingTasks++
		return this.urlUtils.getFileDataFromQueryParam().then((data: NameDataPair[]) => {
			if (data.length > 0) {
				this.tryLoadingFiles(data)
			} else {
				this.tryLoadingSampleFiles()
			}
		}, this.tryLoadingSampleFiles.bind(this))
	}

	public tryLoadingSampleFiles() {
		if (this.urlUtils.getParam("file")) {
			/*this.dialogService.showErrorDialog(
				"One or more files from the given file URL parameter could not be loaded. Loading sample files instead."
			)*/
		}
		this.tryLoadingFiles([
            { name: "sample1.json", data: require("./assets/sample1.json") },
            { name: "sample2.json", data: require("./assets/sample2.json") }
        ]);
    }
    
    private tryLoadingFiles(values: NameDataPair[]) {
        this.codeChartaService
        .loadFiles(
            values
        )
        .then(() => this.viewModel.numberOfLoadingTasks--)
        .catch(e => {
			this.viewModel.numberOfLoadingTasks--
			console.error(e);
            this.printErrors(e)
        })
    }

	public printErrors(errors: Object) {
		//this.dialogService.showErrorDialog(JSON.stringify(errors, null, "\t"))
	}

	private subscribeToLoadingEvents($rootScope: angular.IRootScopeService) {
		$rootScope.$on("add-loading-task", () => {
			this.viewModel.numberOfLoadingTasks++
		})

		$rootScope.$on("remove-loading-task", () => {
			this.viewModel.numberOfLoadingTasks--
		})
	}
}

export const codeChartaComponent = {
	selector: "codeChartaComponent",
	template: require("./codeCharta.component.html"),
	controller: CodeChartaController
}
