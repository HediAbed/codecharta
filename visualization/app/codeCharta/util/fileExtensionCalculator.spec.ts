import { FileExtensionCalculator, FileExtensionDistribution } from "./fileExtensionCalculator"
import { CodeMapNode } from "../codeCharta.model"
import { VALID_NODE_WITH_PATH_AND_EXTENSION } from "./dataMocks"

describe("FileExtensionCalculator", () => {
	let map: CodeMapNode

	beforeEach(() => {
		map = VALID_NODE_WITH_PATH_AND_EXTENSION
	})

	describe("getFileExtensionDistribution", () => {
		it("should get correct distribution of file-extensions with for metric", () => {
			const metrics = ["RLOC", "MCC"]
			const result: FileExtensionDistribution[] = FileExtensionCalculator.getFileExtensionDistribution(map, metrics)
			const expected: FileExtensionDistribution[] = [
				{
					metric: "RLOC",
					distribution: {
						jpg: 130,
						java: 162,
						json: 70,
						none: 15
					}
				},
				{
					metric: "MCC",
					distribution: {
						jpg: 101,
						java: 47,
						json: 10,
						none: 33
					}
				}
			]
			expect(result).toEqual(expected)
		})
	})

	describe("estimateFileExtension", () => {
		it("should return correct lower-cased file extension", () => {
			const fileName: string = "fileName.JAVA"
			const result: string = FileExtensionCalculator["estimateFileExtension"](fileName)
			expect(result).toEqual("java")
		})

		it("should return correct file extension when filename contains multiple points", () => {
			const fileName: string = "prefix.name.suffix.json"
			const result: string = FileExtensionCalculator["estimateFileExtension"](fileName)
			expect(result).toEqual("json")
		})

		it("should return 'none' as extension, as there does not exist any", () => {
			const fileName: string = "name_without_extension"
			const result: string = FileExtensionCalculator["estimateFileExtension"](fileName)
			expect(result).toEqual("none")
		})
	})
})
