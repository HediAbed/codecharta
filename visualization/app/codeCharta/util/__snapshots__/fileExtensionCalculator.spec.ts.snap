// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileExtensionCalculator getFileExtensionDistribution should get correct relative distribution of file-extensions with for metric 1`] = `
Object {
  "RLOC": Array [
    Object {
      "absoluteMetricValue": 162,
      "color": null,
      "fileExtension": "java",
      "relativeMetricValue": 42.97082228116711,
    },
    Object {
      "absoluteMetricValue": 130,
      "color": null,
      "fileExtension": "jpg",
      "relativeMetricValue": 34.48275862068966,
    },
    Object {
      "absoluteMetricValue": 70,
      "color": null,
      "fileExtension": "json",
      "relativeMetricValue": 18.56763925729443,
    },
    Object {
      "absoluteMetricValue": 15,
      "color": null,
      "fileExtension": "None",
      "relativeMetricValue": 3.978779840848806,
    },
  ],
}
`;
